sw/device/lib/base/status.c

```diff
@@ -0,0 +1,128 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/status.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+const uint16_t status_module_id = 0;
+
+static const char *basename(const char *file) {
+  const char *f = file;
+  while (*f)
+    ++f;  // end of string
+  while (f > file && f[-1] != '/')
+    --f;  // last filename path component
+  return f;
+}
+
+status_t status_create(absl_status_t code, uint16_t module_id, const char *file,
+                       int32_t arg) {
+  if (code == kOk) {
+    if (arg >= 0) {
+      return (status_t){.value = arg};
+    } else {
+      // If you find yourself here, then someone returned a OK_STATUS
+      // a negative value.
+      arg = __LINE__;
+    }
+  }
+  /**
+   * Our status(error) codes are arranged as a packed bitfield:
+   *
+   * 32  31      26      21      16             5       0
+   *  +---+-------+-------+-------+-------------+-------+
+   *  |   |   15 bit              | 11 bit      | 5 bit |
+   *  | s |   Module Identifier   | Line Number | code  |
+   *  +---+-------+-------+-------+-------------+-------+
+   *
+   * The sign bit is set on all not-Ok statuses, thus proviging a covenient
+   * overloaded return value from functions that may return an error.
+   */
+  if (module_id == 0) {
+    // First three characters of the filename.
+    const char *f = basename(file);
+    module_id = MAKE_MODULE_ID(f[0], f[1], f[2]);
+  }
+  return (status_t){.error = {
+                        .code = code,
+                        .arg = arg,
+                        .module_id = module_id,
+                        .is_error = 1,
+                    }};
+}
+
+const char *status_codes[] = {
+    "Ok",
+    "Cancelled",
+    "Unknown",
+    "InvalidArgument",
+    "DeadlineExceeded",
+    "NotFound",
+    "AlreadyExists",
+    "PermissionDenied",
+    "ResourceExhausted",
+    "FailedPrecondition",
```

Comment:
What's the difference between `InvalidArgument` and `FailedPrecondition`? Also wondering about `Cancelled` vs `Aborted`. Or are these from a a standard list somewhere?